File: trace-complete-pokemon.js
#!/usr/bin/env node

// Complete trace of PRNG calls for first Pokemon including EVs/IVs

const {PRNG} = require('./../pokemon-showdown-ts/dist/sim/prng');
const Dex = require('./../pokemon-showdown-ts/dist/sim/dex').Dex;

const dex = Dex.mod('gen9');
const prng = new PRNG([0, 0, 0, 1]);

let allSpecies = Object.values(dex.species.all());
allSpecies.sort((a, b) => a.name.localeCompare(b.name));

const allMoves = Object.values(dex.moves.all()).map(m => m.id);
const allItems = Object.values(dex.items.all()).map(i => i.id);
const allNatures = Object.values(dex.natures.all()).map(n => n.id);

let callCount = 0;
const callLog = [];

// Log every PRNG call
const logCall = (method, args, result) => {
    callCount++;
    const entry = `[${callCount}] ${method}(${args}) = ${result}`;
    callLog.push(entry);
    console.log(entry);
};

const originalRandom = prng.random.bind(prng);
const originalSample = prng.sample.bind(prng);
const originalRandomChance = prng.randomChance.bind(prng);

prng.random = function(...args) {
    const result = originalRandom(...args);
    logCall('random', args.join(','), result);
    return result;
};

prng.sample = function(arr) {
    const result = originalSample(arr);
    const displayResult = typeof result === 'object' && result.name ? result.name : result;
    logCall('sample', `arr[${arr.length}]`, displayResult);
    return result;
};

prng.randomChance = function(num, denom) {
    const result = originalRandomChance(num, denom);
    logCall('randomChance', `${num},${denom}`, result);
    return result;
};

console.log('=== COMPLETE POKEMON #1 GENERATION ===\n');

// Generate EVs function
function generateRandomEVs(prng) {
    const evs = [0, 0, 0, 0, 0, 0];
    let totalEVs = 0;

    while (totalEVs < 510) {
        const availableStats = [];
        for (let i = 0; i < evs.length; i++) {
            if (evs[i] < 252) {
                availableStats.push(i);
            }
        }

        if (availableStats.length === 0) break;

        const statIdx = prng.sample(availableStats);
        const amount = Math.min(
            prng.random(1, 5),
            252 - evs[statIdx],
            510 - totalEVs
        );
        evs[statIdx] += amount;
        totalEVs += amount;
    }

    return {
        hp: evs[0],
        atk: evs[1],
        def: evs[2],
        spa: evs[3],
        spd: evs[4],
        spe: evs[5],
    };
}

// 1. Species
const species = prng.sample(allSpecies);
console.log(`\n=> Species: ${species.name}\n`);

// 2. Level
const level = prng.random(50, 101);
console.log(`=> Level: ${level}\n`);

// 3. Item
const item = prng.sample(allItems) || '';
console.log(`=> Item: ${item}\n`);

// 4. Nature
const nature = prng.sample(allNatures);
console.log(`=> Nature: ${nature}\n`);

// 5. Gender
let gender = '';
if (species.genderRatio && species.genderRatio.M > 0 && species.genderRatio.F > 0) {
    gender = prng.randomChance(1, 2) ? 'M' : 'F';
}
console.log(`=> Gender: ${gender || 'N'}\n`);

// 6. Moves
console.log('=> Moves:');
const moves = [];
for (let j = 0; j < 4; j++) {
    const move = prng.sample(allMoves);
    if (!moves.includes(move)) {
        moves.push(move);
    }
}
while (moves.length < 4) {
    moves.push('tackle');
}
console.log(`   ${moves.join(', ')}\n`);

// 7. EVs
console.log('=> EVs:');
const evs = generateRandomEVs(prng);
console.log(`   ${JSON.stringify(evs)}\n`);

// 8. IVs
console.log('=> IVs:');
const ivs = {
    hp: prng.random(0, 32),
    atk: prng.random(0, 32),
    def: prng.random(0, 32),
    spa: prng.random(0, 32),
    spd: prng.random(0, 32),
    spe: prng.random(0, 32),
};
console.log(`   ${JSON.stringify(ivs)}\n`);

console.log(`\n=== TOTAL PRNG CALLS: ${callCount} ===\n`);

// Now test if the next call matches what Rust should get for Pokemon #2
console.log('=== NEXT CALL (for Pokemon #2 species selection) ===');
const usedSpecies = [species.name];
const availableSpecies2 = allSpecies.filter(s => !usedSpecies.includes(s.name));
const species2 = prng.sample(availableSpecies2);
console.log(`Pokemon #2 should be: ${species2.name}`);

File: examples/trace_complete_pokemon.rs
/// Trace PRNG calls for first complete Pokemon in Rust

use pokemon_showdown::{Dex, PRNG, PRNGSeed};
use pokemon_showdown::dex_data::{StatsTable, Gender};

fn main() {
    let dex = Dex::load_default().expect("Failed to load dex");
    let mut prng = PRNG::new(Some(PRNGSeed::Gen5([0, 0, 0, 1])));

    println!("=== COMPLETE POKEMON #1 GENERATION ===\n");

    let mut all_species: Vec<_> = dex.species.values().collect();
    all_species.sort_by_key(|s| &s.name);

    let mut all_moves: Vec<_> = dex.moves.keys().map(|id| id.as_str()).collect();
    all_moves.sort();

    let mut all_items: Vec<_> = dex.items.keys().map(|id| id.as_str()).collect();
    all_items.sort();

    let mut all_natures: Vec<_> = dex.natures.keys().map(|id| id.as_str()).collect();
    all_natures.sort();

    let mut call_count = 0;

    // Wrap PRNG to count calls
    macro_rules! count_call {
        ($method:expr, $call:expr) => {{
            call_count += 1;
            let result = $call;
            println!("[{}] {}", call_count, $method);
            result
        }};
    }

    // 1. Species
    let idx = count_call!("random_range(0, 1515)", prng.random_range(0, all_species.len() as i32));
    let species = all_species[idx as usize];
    println!("=> Species: {}\n", species.name);

    // 2. Level
    let level = count_call!("random_range(50, 101)", prng.random_range(50, 101));
    println!("=> Level: {}\n", level);

    // 3. Item
    let idx = count_call!("random_range(0, 583)", prng.random_range(0, all_items.len() as i32));
    let item = all_items[idx as usize];
    println!("=> Item: {}\n", item);

    // 4. Nature
    let idx = count_call!("random_range(0, 25)", prng.random_range(0, all_natures.len() as i32));
    let nature = all_natures[idx as usize];
    println!("=> Nature: {}\n", nature);

    // 5. Gender
    let gender = if let Some(ref ratio) = species.gender_ratio {
        if ratio.m > 0.0 && ratio.f > 0.0 {
            if count_call!("random_chance(1, 2)", prng.random_chance(1, 2)) {
                Gender::Male
            } else {
                Gender::Female
            }
        } else if ratio.m > 0.0 {
            Gender::Male
        } else if ratio.f > 0.0 {
            Gender::Female
        } else {
            Gender::None
        }
    } else if let Some(ref gender_str) = species.gender {
        // No gender_ratio, but gender field is set (M, F, or N)
        match gender_str.as_str() {
            "M" => Gender::Male,
            "F" => Gender::Female,
            "N" => Gender::None,
            _ => {
                // Default to 50/50 like JavaScript
                if count_call!("random_chance(1, 2)", prng.random_chance(1, 2)) {
                    Gender::Male
                } else {
                    Gender::Female
                }
            }
        }
    } else {
        // No gender_ratio and no gender field - default to 50/50 like JavaScript
        if count_call!("random_chance(1, 2)", prng.random_chance(1, 2)) {
            Gender::Male
        } else {
            Gender::Female
        }
    };
    println!("=> Gender: {:?}\n", gender);

    // 6. Moves
    println!("=> Moves:");
    let mut moves = Vec::new();
    for _ in 0..4 {
        let idx = count_call!("random_range(0, 954)", prng.random_range(0, all_moves.len() as i32));
        let move_name = all_moves[idx as usize].to_string();
        if !moves.contains(&move_name) {
            moves.push(move_name);
        }
    }
    while moves.len() < 4 {
        moves.push("tackle".to_string());
    }
    println!("   {:?}\n", moves);

    // 7. EVs
    println!("=> EVs:");
    let mut evs = [0i32; 6];
    let mut total_evs = 0;
    while total_evs < 510 {
        let mut available_stats = Vec::new();
        for (i, &ev) in evs.iter().enumerate() {
            if ev < 252 {
                available_stats.push(i);
            }
        }
        if available_stats.is_empty() {
            break;
        }

        let idx = count_call!("random_range(0, available)", prng.random_range(0, available_stats.len() as i32));
        let stat_idx = available_stats[idx as usize];

        let amount = count_call!("random_range(1, 5)", prng.random_range(1, 5))
            .min(252 - evs[stat_idx])
            .min(510 - total_evs);
        evs[stat_idx] += amount;
        total_evs += amount;
    }
    let evs_table = StatsTable::new(evs[0], evs[1], evs[2], evs[3], evs[4], evs[5]);
    println!("   hp:{} atk:{} def:{} spa:{} spd:{} spe:{}\n", evs[0], evs[1], evs[2], evs[3], evs[4], evs[5]);

    // 8. IVs
    println!("=> IVs:");
    let iv_hp = count_call!("random_range(0, 32)", prng.random_range(0, 32));
    let iv_atk = count_call!("random_range(0, 32)", prng.random_range(0, 32));
    let iv_def = count_call!("random_range(0, 32)", prng.random_range(0, 32));
    let iv_spa = count_call!("random_range(0, 32)", prng.random_range(0, 32));
    let iv_spd = count_call!("random_range(0, 32)", prng.random_range(0, 32));
    let iv_spe = count_call!("random_range(0, 32)", prng.random_range(0, 32));
    println!("   hp:{} atk:{} def:{} spa:{} spd:{} spe:{}\n", iv_hp, iv_atk, iv_def, iv_spa, iv_spd, iv_spe);

    println!("\n=== TOTAL PRNG CALLS: {} ===\n", call_count);

    // Next call for Pokemon #2
    println!("=== NEXT CALL (for Pokemon #2 species selection) ===");
    let used_species = vec![species.name.as_str()];
    let available_species2: Vec<_> = all_species.iter()
        .filter(|s| !used_species.contains(&s.name.as_str()))
        .copied()
        .collect();

    call_count += 1;
    let idx2 = prng.random_range(0, available_species2.len() as i32);
    println!("[{}] random_range(0, {})", call_count, available_species2.len());
    println!("Pokemon #2 should be: {}", available_species2[idx2 as usize].name);
}
